src/rush/limiters/gcra.py

```diff
@@ -24,62 +30,82 @@ def rate_limit(
         # The increment uses the emission interval to find out how much time
         # quantity should have been issued over
         increment: datetime.timedelta = emission_interval * quantity
+
         now = self.store.current_time()
         data = self.store.get(key)
-        # tat is short for theoretical arrival time, we store this as
-        # "time" on our limit data
-        tat = getattr(data, "time", None) or now
-        new_tat = max(now, tat) + increment
-        # The theoretical arrival time defines the end-period (in the future)
-        # of our bucket. We want to find, however, which is later - our tat or
-        # now. Once we have that, we can find the earliest point in our bucket
-        # which is the time we would start allowing new requests.
-        allow_at: datetime.datetime = new_tat - delay_variation_tolerance
-        # Okay, we have our tat, the time we would have started allowing new
-        # requests, and present time so let's figure out the difference from
-        # our earliest point in time and now.
-        distance_from_start_of_bucket = now - allow_at
-        # Now that we know how far we are from the start, let's find out how
-        # much we have remaining in our quota.
-        remaining = math.floor(
-            (distance_from_start_of_bucket / emission_interval) + 0.5
-        )
-        # We also need to calculate the next reset_after for the user
-        reset_after = tat - now
-        if reset_after.total_seconds() == 0:
-            reset_after = datetime.timedelta(seconds=-1)
-
-        if remaining < 1:
-            # It's possible that distance_from_start_of_bucket is negative so
-            # it is then also possible that there are fewer than 0 remaining
-            # requests available. In that case, we're going to be ratelimited.
-            remaining = 0
-            limited = True
-            retry_after = emission_interval - distance_from_start_of_bucket
-            new_time = tat
-        else:
-            limited = False
-            retry_after = datetime.timedelta(seconds=-1)
-            new_time = new_tat
-
-        used = rate.limit - remaining
-        if data is not None:
-            limitdata = data.copy_with(
-                used=used, remaining=remaining, time=new_time
-            )
-        else:
-            limitdata = limit_data.LimitData(
-                used=used, remaining=remaining, created_at=now, time=new_time
+        attempts = 0
+        while attempts < MAX_CAS_ATTEMPTS:
+            # tat is short for theoretical arrival time, we store this as
+            # "time" on our limit data
+            tat = getattr(data, "time", None) or now
```

Comment:
I'd prefer to make this a separate method

Reply:
Do you mean have a function to check whether to retry which takes the number of attempts and returns a bool? 

Or that you'd prefer to have the contents of the while loop split out into a separate function?

Reply:
The latter

Reply:
Done. As noted in the commit message, the way I've done this does mean that we recalculate `emission_interval` and `delay_variation_tolerance` an extra time for each CAS failure. I suspect this isn't the end of the world though as it's only only division and one multiplication and it saves us passing two additional arguments to `_gcra` which seems much neater.

Reply:
Apologies for squashing the commit just as you reviewed! Makes it a bit tricky to see the further changes I made to this bit of the code.